### ISSUE TITLE
Proposal: make IO auto-cancelable on async boundaries

### ISSUE THREAD

**Issue by alexandru** [typelevel/cats-effect#308]
Given that our type classes now rely on auto-cancelable behavior, having `IO` to not be auto-cancelable makes it less optimal than implementation that are auto-cancelable.

It's not ideal, but at this point polymorphic code has no choice but to rely on `bracket` for resource acquisition and release. Some encodings become invalid, like the older Monix `Iterant` encoding, which went through a painful transition, but that in absence of an alternative needs to be done.

So we have two problems:

1. by having `IO` to not be auto-cancelable, people writing polymorphic code that's tested with `IO` (the default) will get the wrong impression about behavior
2. auto-cancelable implementations will cancel in some cases sooner, avoiding the allocation of resources that `IO` will allocate; for example `task.start.flatMap(_.cancel)` should in general not execute the source (assuming `cancel` happens before the task gets a chance to start), yet with `IO` the source gets executed every single time

Clearly auto-cancelable behavior is now the default. In the future if we'll ever want to bring back this behavior, it has to be opt-in, like in my #242 / #262 proposal for a `Continual` type class.

I was thinking about this as part of #267 because for code like this:

```scala
val task = IO.unit.bracket(_ => IO.never)(_ => IO.sleep(3.seconds))

task.start.flatMap(_.cancel)
```

Depending on the race condition (`cancel` executes before or after the source starts);

1. this operation takes 3 seconds, if the source managed to be executed and thus the finalizer was installed
2. this operation is instant, if `cancel` was executed before the source was started (due to the auto-forking that we are doing) and what happens is that the runtime still executes the source and that finalizer, but you won't observe it on `cancel` (if `cancel` happens first, then the internal cancelation flag is set and all incoming finalizers get executed immediately afterwards)

If `IO` would be auto-cancelable, then in many such cases the source won't even get executed, which is more optimal. Of course, this doesn't entirely fix the race condition, as the race condition in this case isn't easily fixable (e.g. in some cases the execution duration will be instance, instead of 3 seconds, even if the source gets executed), but such instances will be rare.

Steps to accomplish this:

1. check the internal `IOPlatform.isCanceled` flag after each `IO.Async` state was evaluated, before proceeding with its bind continuation
2. check this flag just after the automated forks we do in `start` or `race`, but before executing the source ... so for example in `start` we would first fork, then check the flag, then execute the source
3. we need to get rid of `onCancelRaiseError`; at this point this operation becomes dangerous

I can do a PR for it.

**Comment by SystemFw** [comment]
Sounds great :)
I only have a question about 3, how do you get the equivalent functionality? `bracketCase` + pattern matching on `Cancelled`?

**Comment by alexandru** [comment]
Yes, `bracketCase` + pattern matching on `Canceled`, or you can use `guaranteeCase` in case you're not dealing with a resource.

**Comment by mpilquist** [comment]
I like it too.
